src/library_syscall.js

```diff
@@ -1680,13 +1660,13 @@ for (var x in SyscallsLibrary) {
 #if USE_PTHREADS
 // emscripten_syscall is a switch over all compiled-in syscalls, used for proxying to the main thread
 var switcher =
-  'function(which, varargs) {\n' +
+  'function(which) {\n' +
   '  switch (which) {\n';
 DEFAULT_LIBRARY_FUNCS_TO_INCLUDE.forEach(function(func) {
   var m = /^__syscall(\d+)$/.exec(func);
   if (!m) return;
   var which = +m[1];
-  switcher += '    case ' + which + ': return ___syscall' + which + '(which, varargs);\n';
+  switcher += '    case ' + which + ': return ___syscall' + which + '.apply(arguments.slice(1));\n';
```

Comment:
this is going to cause slowdowns. maybe offset by the other benefits here though. but i believe there is a better way, you can find the function and see how many arguments it has using `func.length`.